README.md

```diff
@@ -19,44 +19,241 @@ ICAT API to interface with the Data Gateway
+
+```bash
+sudo yum install @development zlib-devel bzip2 bzip2-devel readline-devel sqlite \
+sqlite-devel openssl-devel xz xz-devel libffi-devel findutils
+```
+
+To make use of `pyenv`, let's install different versions of Python onto the system. In
+production, DataGateway API uses Python 3.6, so this should definitely be part a
+development environment for this repo. This stage might take some time as each Python
+version needs to be downloaded and built individually:
+
+```bash
+pyenv install 3.6.8
+pyenv install 3.7.7
+pyenv install 3.8.2
+```
+
+To verify the installation commands worked:
+
+```bash
+python3.6 --version
+python3.7 --version
+python3.8 --version
+```
+
+These Python versions need to be made available to local version of the repository. They
+will used during the Nox sessions, explained further down this file. Executing the
+following command will create a `.python-version` file inside the repo (this file is
+currently listed in `.gitignore`):
+
+```bash
+pyenv local 3.6.8 3.7.7 3.8.2
+```
+
+### Poetry (API Dependency Management)
+To maintain records of the API's dependencies,
+[Poetry](https://github.com/python-poetry/poetry) is used. To install, use the following
+command:
 
-## Setup and running the API   
-The database connection needs to be set up first. This is set in config.json, an example
-config file called `config.json.example` is provided.
+```bash
+curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -
+```
+
+The installation requires `~/.poetry/env` to be refreshed for changes to be applied.
+Open a new terminal or execute the following command to ensure the installation is
+completed smoothly:
+
+```bash
+source ~/.poetry/env
+```
+
+The dependencies for this repo are stored in `pyproject.toml`, with a more detailed
+version of this data in `poetry.lock`. The lock file is used to maintain the exact
+versions of dependencies from system to system. To install the dependencies, execute the
+following command (add `--no-dev` if you don't want the dev dependencies):
+
+```bash
+poetry install
+```
+
+To add a dependency to Poetry, run the following command (add `--dev` if it's a
+development related dependency). The
+[official docs](https://python-poetry.org/docs/cli/#add) give good detail regarding the
+intricacies of this command:
+
+```bash
+poetry add [PACKAGE-NAME]
+```
+
+### Nox (Automated Testing & Other Code Changes)
+When developing new features for the API, there are a number of Nox sessions that can be
+used to lint/format/test the code in the included `noxfile.py`. To install Nox, use Pip
+as shown below. Nox is not listed as a Poetry dependency because this has the potential
+to cause issues if Nox was executed inside Poetry (see
+[here](https://medium.com/@cjolowicz/nox-is-a-part-of-your-global-developer-environment-like-poetry-pre-commit-pyenv-or-pipx-1cdeba9198bd)
+for more detailed reasoning):
+
+```bash
+pip install --user --upgrade nox
```

Comment:
I will add a note somewhere along the lines of:
"if you have issues, remove the `--user` option".

Reply:
So I've just re-read this, and I've realised you've experienced this error because you installed `nox` and `pre-commit` in a virtual environment. When installing Python packages in a virtualenv, the `--user` shouldn't be used (you're only affecting that isolated environment, not the system Python). I put a link in the start of that section to explain why these packages aren't installed in a virtual environment, as explained by the author of the hypermodern python guide. In an ideal world where all Python repos we worked on were set up in the same way, this would make life a lot easier.

To explain a bit more explicitly, I've added a sentence at the end of that paragraph to say that option isn't needed if you choose to install the packages in a virtual environment. As a result, I see this as resolved.

Reply:
I can see why I was getting the error when I was running the command with the `--user` flag. However, I tried installing `nox` outside the virtual environment that is created by Poetry but when I tried to run the `nox` command I got an `nox: command not found` error. Same thing happened when I tried to run `pre-commit install`. I could see that both packages were installed when I ran `pip list` tho. 

**Solution**:
My Python 3.8 is installed in `C:\Python38` and when I ran `which python`, it was pointing to that directory, however I found that pip did not install the packages in this directory but in `C:\Users\user\AppData\Roaming\Python\Python38` instead. Because of this I tried adding the following two paths to the `PATH` system variable and restarted my machine, and that fixed the issue:
`C:\Users\user\AppData\Roaming\Python\Python38\Scripts\`
`C:\Users\user\AppData\Roaming\Python\Python38\`

It might be worth adding a note about this in the README.

Reply:
Today I've been working on documentation in a different branch. For simplicity and to ensure I don't miss it during a merge conflict, I've added your suggested change to a commit on that branch, see https://github.com/ral-facilities/datagateway-api/commit/f2f7f55eff5ea86202471ea7b709d24747d7a606. I see this as resolved.